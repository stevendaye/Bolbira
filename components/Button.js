/*#### Button Component #### */
import React from "react";

const Button = ({ className = "", onClick, children, type }) =>
  <button
    type = {type}
    className = {className}
    onClick = {onClick}
  >
    {children}
  </button>

export default Button;
