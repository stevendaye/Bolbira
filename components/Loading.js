/* Creating a Loading component */
import React from "react";
import PropTypes from "prop-types";

const Loading = ({ error }) =>
  <div className = { error? "hide" : "loading"}>
    Loading...
  </div>

// Checking Loading Props type;
Loading.propTypes = {
  error: PropTypes.bool.isRequired
};

export default Loading;
