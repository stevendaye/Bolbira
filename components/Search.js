/* ## Search Component ## */
import React from "react";
import Button from "./Button";

const Search = ({ value, onChange, placeholder, type, onSubmit, onClick, children }) =>
  <div className = "search-box-wrapper">
    <form onSubmit = {onSubmit} >
      <input
        className = "search-box"
        value= {value}
        type = {type}
        placeholder = {placeholder}
        onChange = {onChange}
      />
      <Button
        type = "submit"
        className = "search-button"
        onClick = {onClick}
      >
        {children}
      </Button>
    </form>
  </div>

export default Search;