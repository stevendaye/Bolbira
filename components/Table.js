/* List Table Styles */
import React from "react";
import Button from "./Button";
import PropTypes from "prop-types";

const Table = ({ list, onDismiss }) => (
  <div className = "search-result">

      <div className = "sort-points">...</div>
      <div className = "clear"></div>

    {list.map( item =>
      <section key = {item.objectID} style = { (!item.title || !item.url) ? {display: "none"} : {display: "block"}}>
        <h3><a href= {item.url}>{item.title}</a></h3>
        <div className = "clear"></div>
        <i>{item.author}</i>
        <p><a href= {item.url}>{item.url}</a></p>
        <hr/>
        <p>{item.title}<a href= {item.url}><em className = "read-more">...Read More...</em></a></p>
        <span>Comments: {item.num_comments}</span>
        <span>Points: {item.points}</span>
        <Button
          type = "button"
          className = "rm-btn"
          onClick = {() => {onDismiss(item.objectID)}}
        >
          Dismiss
        </Button>
        <span className = "search-opt">...</span>
      </section>
    )}
  </div>
);

// Checking the types of the Table props
Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.number.isRequired,
      title: PropTypes.string,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  ).isRequired,
  onDismiss: PropTypes.func.isRequired
};

export default Table;