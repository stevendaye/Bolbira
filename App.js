/* ## bolbir main file: App.js ## */
import React, { Component } from "react";
import Search from "./components/Search";
import Table from "./components/Table";
import Loading from "./components/Loading";
import Err from "./components/Error";
import Footer from "./components/Footer";
import "./src/css/app.css"
import Button from "./components/Button";
import axios from "axios";

const PATH_BASE = "http://hn.algolia.com/api/v1";
const PATH_SEARCH = "/search";
const PARAM_SEARCH = "query="
const DEFAULT_SEARCH = "redux";
const PARAM_PAGE = "page=";
const PARAM_HPP = "hitsPerPage=";
const DEFAULT_HPP = "15";

class App extends Component {
  // Using ES6 Class Field to prevent the component from updating the state when it's unmounted
  _isMounted = false;

  constructor (props) {
    super(props);

    this.state = {
      results: null,
      searchTerm: DEFAULT_SEARCH,
      error: null,
      isLoading: false,
      isLoadingMore: false,
      searchKey: ""
    };

    this.loadMoreTopStories = this.loadMoreTopStories.bind(this);
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.fecthSearchTopStories = this.fecthSearchTopStories.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.notify = this.notify.bind(this);
  }

  // Defining core coponent methods
  fecthSearchTopStories(searchTerm, page = 0) {
    this.setState({ isLoading: true });
    axios(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(result => this._isMounted && this.setSearchTopStories(result.data))
      .catch(error => this._isMounted && this.setState({ error }));
  }
  setSearchTopStories (result) {
    const { hits, page } = result;
    this.setState( prevState => {
      const { results, searchKey } = prevState;
      const old_hits = (results && results[searchKey]) ? results[searchKey].hits : [];
      const updatedHits = [ ...old_hits, ...hits];
      return {
        results: {
          ...results,
          [searchKey]: {
            hits: updatedHits,
            page
          }
        },
        isLoading: false
      }
    });
  } 
  onSearchSubmit (e) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });
    this.needsToSearchTopStories(searchTerm) && this.fecthSearchTopStories(searchTerm);
    e.preventDefault();
  }
  needsToSearchTopStories (searchTerm) {
    const { results } = this.state;
    return !results[searchTerm];
  }

  onDismiss (id) {
    this.setState( prevState => {
      const { results, searchKey } = prevState;
      const {hits, page} = results[searchKey];
      const isNotID = item => item.objectID !== id;
      const updatedHits = hits.filter(isNotID);
      return {
        results: {
          ...results,
          [searchKey]: {
            hits: updatedHits,
            page
          }
        }
      }
    })
  }
  onSearchChange (e) {
    this.setState({ searchTerm: e.target.value });
  }
  notify () {
    console.log("Search Term Sent.");
  }
  loadMoreTopStories () {
    this.setState({ isLoadingMore: true });
  }

  // Using core component licycles methods
  componentDidMount () {
    this._isMounted = true;
    const { searchTerm } = this.state;
    this.fecthSearchTopStories(searchTerm);
  }
  componentWillUnmount () {
    this._isMounted = false;
  }

  render () {
    const hello = "Hello I am bolbir. Your Friendly Search App";

    const {
      results,
      searchTerm,
      error,
      isLoading,
      isLoadingMore,
      searchKey
    } = this.state;

    const page = (
      results &&
      results[results] &&
      results[searchKey].page
    ) || 0; // In case there is no results, so the page number is 0;

    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
    ) || []; // In case there is no list result, it returns an empty array. With this, no more need to do a conditional rendering;
    
    return (
      <div className = "app_wrapper">
            
        <header>
          <div className = "up-front">
            <div className = "app-name">
              <h1>bolbir</h1>
            </div>
            <nav>
              <ul>
                <li><a href="#">Sign in</a></li>
                <li><a href="#">Popular</a></li>
                <li><a href="#">Pin</a></li>
              </ul>
            </nav>
            <h2>{hello}</h2>
            <Search
              value = {searchTerm}
              placeholder = "Type something you are interested in!"
              type = "search"
              onChange = {this.onSearchChange}
              onSubmit = {this.onSearchSubmit}
              onClick = {this.notify}
            >
              Search
            </Search>
          </div>
          <div>
            <h4>
              <small><a href = "#">I am looking for tech related contents</a></small>&nbsp;&nbsp;&nbsp;<span>|
              </span>&nbsp;&nbsp;<small><a href = "#">Person of Interest</a></small>
            </h4>  
          </div>
        </header>

        <TableWithError
          error = {error}
          list = {list}
          onDismiss = {this.onDismiss}
        />
        <ButtonWithLoading
          type = "button"
          className = "more-btn"
          error = {error}
          isLoading = {isLoading}
          isLoadingMore = {isLoadingMore}
          onClick = {() => { this.loadMoreTopStories() || this.fecthSearchTopStories(searchKey, page + 1) }}
        >
          +
        </ButtonWithLoading>
        
        <Footer/>
      </div>
    );
  }
}

// Creating a HOC for the Error and Table Components
const withError = Component => ({ error, ...rest }) =>
  error
    ? <Err/>
    : <Component {...rest} />

const TableWithError = withError(Table);

// Creating a HOC for the Loading and More Button Components
const withLoading = Component => ({ isLoading, error, isLoadingMore, ...rest }) =>
  isLoading
    ? <Loading
        error = {error}
        isLoadingMore = {isLoadingMore}
      />
    : <div className = "more-btn-wrapper">
        <Component {...rest} />
      </div>

const ButtonWithLoading = withLoading(Button);

export default App;