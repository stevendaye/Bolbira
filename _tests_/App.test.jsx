/* #### Testing our App #### */
import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Enzyme, {shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import App from "../App";
import Table from "../components/Table"
import Search from "../components/Search"
import Loading from "../components/Loading"
import Err from "../components/Error";
import Button from "../components/Button";

Enzyme.configure({ adapter: new Adapter });

// Creating a Test Suit for the App Component
describe ("App", () => {

  // Writing a test case for rendering properly the component
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <App/>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  // Writting Snapshot test for future checks
  test("has a valid snapshot", () => {
    const component = renderer.create(<App/>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

// Creating a test suit for the Search Component
describe("Search", () => {

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Search>
        Search
      </Search>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(
      <Search>
        Search
      </Search>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

// Creating a test suit for the Loading Component
describe("Loading", () => {
  
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Loading/>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(
      <Loading/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

// Creating a test suit for the Error Component
describe("Error", () => {

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Err/>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(<Err/>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

// Creating a test suit for the Button Component
describe("Button", () => {

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Button>
        Click Me
      </Button>, div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  test("has a valid snapshot", () =>{
    const component = renderer.create(
      <Button>
        Click Me
      </Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});

// Creating a test suit for the Table Component
describe("Table", () => {

  const props = {
    list: [
      {title: "React", author: "Clark", num_comments: 12, points: 7, objectID: 0},
      {title: "Redux", author: "Abramov", num_comments: 77, points: 40, objectID: 1},
      {title: "Flux", author: "Simon", num_comments: 144, points: 3, objectID: 2}
    ]
  };

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Table
        {...props}
      />, div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("has three elements", () => {
    const element = shallow(
      <Table
        {...props}
      />
    );
    expect(element.find(".search-result").length).toBe(3);
  });

  test("has a valid snapshot", () => {
    const component = renderer.create(
      <Table
        {...props}
      />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});