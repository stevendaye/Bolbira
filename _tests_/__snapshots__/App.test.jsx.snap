// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="app_wrapper"
>
  <header>
    <div
      className="up-front"
    >
      <div
        className="app-name"
      >
        <h1>
          bolbir
        </h1>
      </div>
      <nav>
        <ul>
          <li>
            <a
              href="#"
            >
              Sign in
            </a>
          </li>
          <li>
            <a
              href="#"
            >
              Popular
            </a>
          </li>
          <li>
            <a
              href="#"
            >
              Pin
            </a>
          </li>
        </ul>
      </nav>
      <h2>
        Hello I am bolbir. Your Friendly Search App
      </h2>
      <div
        className="search-box-wrapper"
      >
        <form
          onSubmit={[Function]}
        >
          <input
            className="search-box"
            onChange={[Function]}
            placeholder="Type something you are interested in!"
            type="search"
            value="redux"
          />
          <button
            className="search-button"
            onClick={[Function]}
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </div>
    <div>
      <h4>
        <small>
          <a
            href="#"
          >
            I am looking for tech related contents
          </a>
        </small>
           
        <span>
          |
        </span>
          
        <small>
          <a
            href="#"
          >
            Person of Interest
          </a>
        </small>
      </h4>
    </div>
  </header>
  <div
    className="search-result"
  >
    <div
      className="sort-points"
    >
      ...
    </div>
    <div
      className="clear"
    />
  </div>
  <div
    className="loading"
  >
    Loading...
  </div>
</div>
`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
>
  Click Me
</button>
`;

exports[`Error has a valid snapshot 1`] = `
<div
  className="error"
>
  Oups, Something went wrong!
  <br />
  Please, check your internet connection or update your date!
</div>
`;

exports[`Loading has a valid snapshot 1`] = `
<div
  className="loading"
>
  Loading...
</div>
`;

exports[`Search has a valid snapshot 1`] = `
<div
  className="search-box-wrapper"
>
  <form>
    <input
      className="search-box"
    />
    <button
      className="search-button"
      type="submit"
    >
      Search
    </button>
  </form>
</div>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="search-result"
>
  <div
    className="sort-points"
  >
    ...
  </div>
  <div
    className="clear"
  />
  <section
    style={
      Object {
        "display": "none",
      }
    }
  >
    <h3>
      <a>
        React
      </a>
    </h3>
    <div
      className="clear"
    />
    <i>
      Clark
    </i>
    <p>
      <a />
    </p>
    <hr />
    <p>
      React
      <a>
        <em
          className="read-more"
        >
          ...Read More...
        </em>
      </a>
    </p>
    <span>
      Comments: 
      12
    </span>
    <span>
      Points: 
      7
    </span>
    <button
      className="rm-btn"
      onClick={[Function]}
      type="button"
    >
      Dismiss
    </button>
    <span
      className="search-opt"
    >
      ...
    </span>
  </section>
  <section
    style={
      Object {
        "display": "none",
      }
    }
  >
    <h3>
      <a>
        Redux
      </a>
    </h3>
    <div
      className="clear"
    />
    <i>
      Abramov
    </i>
    <p>
      <a />
    </p>
    <hr />
    <p>
      Redux
      <a>
        <em
          className="read-more"
        >
          ...Read More...
        </em>
      </a>
    </p>
    <span>
      Comments: 
      77
    </span>
    <span>
      Points: 
      40
    </span>
    <button
      className="rm-btn"
      onClick={[Function]}
      type="button"
    >
      Dismiss
    </button>
    <span
      className="search-opt"
    >
      ...
    </span>
  </section>
  <section
    style={
      Object {
        "display": "none",
      }
    }
  >
    <h3>
      <a>
        Flux
      </a>
    </h3>
    <div
      className="clear"
    />
    <i>
      Simon
    </i>
    <p>
      <a />
    </p>
    <hr />
    <p>
      Flux
      <a>
        <em
          className="read-more"
        >
          ...Read More...
        </em>
      </a>
    </p>
    <span>
      Comments: 
      144
    </span>
    <span>
      Points: 
      3
    </span>
    <button
      className="rm-btn"
      onClick={[Function]}
      type="button"
    >
      Dismiss
    </button>
    <span
      className="search-opt"
    >
      ...
    </span>
  </section>
</div>
`;
